In [4]:
import pandas as pd
from collections import defaultdict
 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

#A case
val = df.Survived.value_counts(normalize=True)
print(val)

#B Case
bVal = df.groupby('Sex')['Survived'].mean()
print(bVal)

#C Case
Q1 = df['Fare'].quantile(0.25)
Q3 = df['Fare'].quantile(0.75)

iqr = Q3 - Q1

max = Q3 + (1.5 * iqr)

print(max)
countMax = 0
map_val = defaultdict(int)

for val in df.itertuples():
    if val.Fare > max:
        countMax += 1
        map_val[val.Pclass] += 1

print("Count High : " , countMax)
print("Map : ", map_val)

#D Case
avg = df.Fare.mean()
median = df.Fare.median()

print("Average : ", avg)
print("Median : ", median)


## E Case
proportion = df["Survived"].value_counts(normalize=True)
print(proportion)

n = 150
not_survived = round(proportion[0] * n)
survived = round(proportion[1] * n)

print(not_survived)
print(survived)

not_survived_people = df[df["Survived"] == 0].sample(
    n = not_survived,
    random_state =42
)

survived_people = df[df["Survived"] == 1].sample(
    n = survived,
    random_state = 42
)


muestra = pd.concat([not_survived_people, survived_people])
print(muestra)

###F Case
print(df.groupby('Survived')['Age'].mean())

Survived
0    0.616162
1    0.383838
Name: proportion, dtype: float64
Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64
65.6344
Count High :  116
Map :  defaultdict(<class 'int'>, {1: 104, 2: 5, 3: 7})
Average :  32.204207968574636
Median :  14.4542
Survived
0    0.616162
1    0.383838
Name: proportion, dtype: float64
92
58
     PassengerId  Survived  Pclass  \
312          313         0       2   
124          125         0       1   
783          784         0       3   
167          168         0       3   
772          773         0       2   
..           ...       ...     ...   
237          238         1       2   
190          191         1       2   
513          514         1       1   
166          167         1       1   
869          870         1       3   

                                                Name     Sex   Age  SibSp  \
312            Lahtinen, Mrs. William (Anna Sylfven)  female  26.0      1   
124                      White, Mr. Perc

Pregunta A (Sumarización Categórica): Usen la función .value_counts(normalize=True) en la columna Survived (0 = Murió, 1 = Sobrevivió). ¿Cuál es la tasa de supervivencia global del barco expresada en porcentaje?

De acuerdo a la funcion, el 38% de las personas del barco sobrevivieron al accidente

Pregunta B : El famoso código marítimo era "mujeres y niños primero". Comprobemos esto matemáticamente. Ejecuten una agrupación por la columna de género y calculen el promedio de la columna Survived. ¿Qué porcentaje exacto de mujeres sobrevivió en contraste con los hombres?

El 74% de mujeres sobrevivieron al accidente en contraste al 18% de hombres

Pregunta C: El director financiero nota que la tarifa máxima (Fare) cobrada fue de más de 500 libras, mientras que el promedio ronda las 32. Sospecha de outliers. Utilicen las herramientas de dispersión en Pandas para confirmarlo:
Calculen el Cuartil 1 (25%) y el Cuartil 3 (75%) de la columna Fare usando el método .quantile([0.25, 0.75]).
Calculen matemáticamente el Rango Intercuartílico (IQR = Q3 - Q1).
Calculen el límite superior aceptable (Q3 + 1.5 * IQR). Escriban una línea de código para filtrar el DataFrame y descubrir exactamente cuántos pasajeros pagaron una tarifa por encima de ese límite matemático. ¿A qué clase (Pclass) pertenecían la mayoría de ellos?

Fueron 116 pasajeros que pagaron una tarifa mayor y 104 de ellos eran de la clase 1

Pregunta D: Calculen la media y la mediana de la columna Fare. Notarán una diferencia enorme entre ambos valores. Matemáticamente, ¿qué significa que la media sea tan superior a la mediana? Si en el futuro utilizamos un algoritmo basado en distancias euclidianas (como K-Nearest Neighbors) sin escalar previamente esta variable, ¿cómo afectará esta asimetría al aprendizaje del modelo?

1. Significa que hay un desbalance de los valores hacia el lado mayor o derecha creando una distribucion sesgada, esto en un futuro podrias crear una asimetria en donde se tomarian valoires muy diferenciales que no relefejan como interactuan los datos entre si, teniendo un peso mayor al que realmente significa

Pregunta E: En la semana 2 vimos que el muestreo aleatorio simple es peligroso. El objetivo es entrenar un modelo que prediga la supervivencia (Survived), pero las clases están desbalanceadas. Escriban el código en Pandas para extraer una muestra de exactamente 150 pasajeros garantizando que la proporción de sobrevivientes y no sobrevivientes en la muestra sea idéntica a la de la base de datos completa. ¿Qué sesgo evitan al hacer esto?

Evitamos tener todos los los valores en no survided y tener un sesgo que nos diga que nadie sobrebibio cuando la realidad, es que un 30% sobrevivieron al accidente, por lo que la tener este tipo de muestro aseguramos estabilida igual de muestrac en el algoritmo

Pregunta F: Si ejecutan df.groupby('Survived')['Age'].mean(), Pandas calculará el promedio de edad de los que vivieron y los que murieron. Sin embargo, por defecto, Pandas ignora los valores NaN al calcular la media. Si resulta que la gran mayoría de las edades faltantes (NaN) pertenecían a pasajeros de 3ra clase que murieron, ¿qué sesgo estadístico estamos introduciendo involuntariamente en el resultado de esa función y cómo afectaría la inferencia de nuestro modelo?. Lo que pasaria es que al momento de hacer las predicciones, el mnodelo podria llegar a inferir que no exixtian pasajeros en la tercera clase, ya que pandas como los ignora, pues no habria la suficiente informacion para clasificar algun posible usuariod entro de esa clase, por lo que el modelo terminaqria sesgado y no podria obtener una respuesta clara, una solucion seria utilizar la media o mediana para cubrir los NaN en las edades de los pasajeros

Pregunta G: Con el cálculo del IQR en la Pregunta C, determinaron que los boletos de 512 libras son outliers matemáticos. En Machine Learning, los valores atípicos pueden representar errores de captura (ruido que aumenta el error irreducible) o casos especiales válidos (señal). Investigando la naturaleza de un barco de lujo, ¿deberíamos eliminar estas filas con .drop() antes de entrenar nuestro modelo? Justifiquen su respuesta arquitectónica.

